### PR TITLE
[docs] various asset concept docs changes

### PR DESCRIPTION
**Pull request by jamiedemaria**
[low priority]

## Summary & Motivation

I was reading through the assets concepts pages to determine what examples need to be changed to use the `deps` API, and I ran across a couple random places that I think could use a little improvement (not related to `deps`). Opening them in a PR here to propose some changes. I'm leaving my reasoning for each change in comments, and please feel free to accept/reject individual changes. I erred on the side of making bigger/more dramatic changes with the thinking that it's easier to say no or scale back changes

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #15396** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/15396" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/15396?utm_source=stack-comment).

File: docs/content/concepts/assets/graph-backed-assets.mdx [hunk 1]
```diff
@@ -58,9 +58,9 @@ def slack_files_table():
 
 ### Defining basic dependencies for graph-backed assets
 
-Similar to with basic software-defined assets, Dagster infers the upstream assets from the names of the arguments to the decorated function.
+Similar to software-defined assets, Dagster infers the upstream assets from the names of the arguments to the decorated function.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/graph-backed-assets.mdx:61]: small grammar fix

File: docs/content/concepts/assets/graph-backed-assets.mdx [hunk 2]
```diff
@@ -58,9 +58,9 @@ def slack_files_table():
...
 
-The example below includes an asset named `middle_asset`. `middle_asset` depends on `upstream_asset`, and `downstream_asset` depends on `middle_asset`:
+The example below includes a graph-backed asset named `middle_asset`. `middle_asset` depends on `upstream_asset`, and `downstream_asset` depends on `middle_asset`:
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/graph-backed-assets.mdx:63]: i felt like being more specific that `middle_asset` was the graph backed asset made it easier to find the important part of this example

File: docs/content/concepts/assets/graph-backed-assets.mdx [hunk 3]
```diff
@@ -93,18 +93,26 @@ def downstream_asset(middle_asset):
 
 ### Graph-backed multi-assets
 
-Using the <PyObject object="graph_multi_asset" decorator />, you can create a combined definition of multiple assets that are computed using the same graph of ops and same upstream assets.
+If you want to write a graph-backed asset that results in the materialization of multiple assets, you can use the <PyObject object="graph_multi_asset" decorator />.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/graph-backed-assets.mdx:96]: I found this section pretty confusing/hard to parse. I boiled it down to two reasons:
* the intro sentence was quite complicated, lots of noun soup with `combined definition of multiple assets that are computed using the same graph of ops and same upstream assets` 
* the example didn't explain what `two_outputs` did, nor did it really need to be a graph-backed multi asset since it was a single op (could have just been a plain multi asset)

So i played with the intro sentence a bit. This read easier to me, but that could just be a personal reading style thing, so very open to feedback.

I also updated the example to do some computation (even though that computation is trivial). This is a personal nit with some of our docs examples, but i feel like we often force the user to imagine/infer what the functions do, which can take away from understanding how the pieces fit together

File: docs/content/concepts/assets/multi-assets.mdx [hunk 4]
```diff
@@ -132,3 +134,23 @@ def my_complex_assets(a, b):
     # d only depends on b
     yield Output(value=b + 1, output_name="d")
 ```
+
+---
+
+## Multi-asset code versions
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/multi-assets.mdx:140]: I moved the multi-asset specific code version docs to the multi-asset concept page because I felt that they didn't fit in the assets concept page. The asset concept page is probably one of the first ones a user looks at after the tutorial, and this particular section introduces two new concepts (multi assets and ops) that are not strictly necessary yet

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 5]
```diff
@@ -25,9 +25,9 @@ A software-defined asset includes the following:
 
 - An <PyObject object="AssetKey" />, which is a handle for referring to the asset.
 - A set of upstream asset keys, which refer to assets that the contents of the software-defined asset are derived from.
-- An [op](/concepts/ops-jobs-graphs/ops), which is a function responsible for computing the contents of the asset from its upstream dependencies.
+- A Python function which is responsible for computing the contents of the asset from its upstream dependencies.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:28]: This is maybe a more controversial change, so if you don't think it makes sense that's totally cool. putting myself in the shoes of a brand new user, I envision my path through the docs to be 1) tutorial, 2) concept pages. Specifically, I would start with the assets concept page. and being presented with the concept of ops right away would be pretty intimidating (thoughts like "ok so i need to go read that concept page first?", "maybe i won't understand assets if i dont understand ops" etc)

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 6]
```diff
@@ -544,22 +528,22 @@ def test_uses_resource() -> None:
     assert result == {"foo": "bar"}
 ```
 
-If you use a context object in your function, you can use <PyObject object="build_op_context" /> to generate the `context` object, because under the hood the function decorated by `@asset` is an op.
+If you use a context object in your function, you can use <PyObject object="build_asset_context" /> to generate the `context` object.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:531]: just updating this section to use the new `build_asset_context` api

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 7]
```diff
@@ -629,15 +613,31 @@ def table1() -> Output[DataFrame]:
...
-from dagster import Output, asset
+from dagster import AssetExecutionContext, asset
 
 
 @asset
-def table1() -> Output[None]:
+def table1(context: AssetExecutionContext) -> None:
     ...  # write out some data to table1
-    return Output(None, metadata={"num_rows": 25})
+    context.add_output_metadata(metadata={"num_rows": 25})
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:622]: not 100% sure this is what we want to recommend, but ti's the approach we are taking in the tutorial for when an asset had no return value, but you want to attach metadata. We should update this to whatever the outcome of this [discussion](https://github.com/dagster-io/dagster/discussions/15392) is

File: docs/content/concepts/assets/asset-observations.mdx [hunk 8]
```diff
@@ -5,7 +5,7 @@ description: Dagster provides functionality to record metadata about assets.
 
 # Asset Observations
 
-An asset observation is an event that records metadata about a given asset. Unlike asset materializations, asset observations do not signify that an asset has been mutated.
+You may wish to record metadata about an asset outside of that asset's materialization function. To do this, you can write another function and log <PyObject object="AssetObservation" /> events. An asset observation is an event that records metadata about a given asset. Unlike asset materializations, asset observations do not signify that an asset has been mutated.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/asset-observations.mdx:8]: i felt like the intro was missing some motivation on why you would use an AssetObservation (rather than just log metadata in the asset materialization function). I still sort of feel that way, but this felt like a small improvement

File: docs/content/concepts/assets/graph-backed-assets.mdx [hunk 9]
```diff
@@ -27,11 +27,11 @@ In this case, when you tell Dagster to materialize the `slack_files_table` asset
 
 ```python file=/concepts/assets/graph_backed_asset.py startafter=start example endbefore=end example
 import pandas as pd
-from dagster import graph_asset, op
+from dagster import OpExecutionContext, graph_asset, op
 
 
 @op(required_resource_keys={"slack"})
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/graph-backed-assets.mdx:33]: make this a Pythonic resource

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 10]
```diff
@@ -86,12 +86,17 @@ from dagster import asset
 
 @asset
 def sugary_cereals() -> None:
-    execute_query("CREATE TABLE sugary_cereals AS SELECT * FROM cereals")
+    execute_query(
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:89]: updated these queries to be a bit more interesting than copying the whole table

File: docs/content/concepts/assets/multi-assets.mdx [hunk 11]
```diff
@@ -22,6 +22,8 @@ Multi-assets may be useful in the following scenarios:
 - A single call to an API results in multiple tables being updated (e.g. Airbyte, Fivetran, dbt).
 - The same in-memory object is used to compute multiple assets.
 
+---
```
[the remark this answers is not in the document]
Comment (erinkcochran87) [reply docs/content/concepts/assets/multi-assets.mdx:25]: I like this a lot. I think we could do a better job of telling folks there are recommended prerequisites for concepts that go beyond surface level

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 12]
```diff
@@ -25,11 +25,11 @@ A software-defined asset includes the following:
...
 
-**Materializing** an asset is the act of running its op and saving the results to persistent storage. You can initiate materializations from [the Dagster UI](/concepts/webserver/ui) or by invoking Python APIs. By default, assets are materialized to pickle files on your local filesystem, but materialization behavior is fully customizable using [I/O managers](/concepts/io-management/io-managers#applying-io-managers-to-assets). It's possible to materialize an asset in multiple storage environments, such as production and staging.
+**Materializing** an asset is the act of running its function and saving the results to persistent storage. You can initiate materializations from [the Dagster UI](/concepts/webserver/ui) or by invoking Python APIs. By default, assets are materialized to pickle files on your local filesystem, but materialization behavior is fully customizable using [I/O managers](/concepts/io-management/io-managers#applying-io-managers-to-assets). It's possible to materialize an asset in multiple storage environments, such as production and staging.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:32]: i did a pass where i tried to remove the word `op` as much as possible. Will mark them as such (with a [1]), and as before, feel free to accept/reject individual changes or the entire `op` replacement

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 13]
```diff
@@ -38,7 +38,7 @@ A software-defined asset includes the following:
...
 | ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
 | <PyObject object="asset" decorator /> | A decorator used to define assets.                                                                                                                                                                                                   |
-| <PyObject object="SourceAsset" />     | A class that describes an asset, but doesn't define how to compute it. <PyObject object="SourceAsset" />s are used to represent assets that other assets or jobs depend on, in settings where they can't be materialized themselves. |
+| <PyObject object="SourceAsset" />     | A class that describes an asset, but doesn't define how to compute it. <PyObject object="SourceAsset" />s are used to represent assets that other assets depend on, in settings where they can't be materialized themselves. |
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:41]: removed "or jobs" since it didn't seem necessary and references a concept the user may not have familiarity with

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 14]
```diff
@@ -63,7 +63,7 @@ def my_asset():
     return [1, 2, 3]
 ```
 
-By default, the name of the decorated function, `my_asset`, is used as the asset key. The decorated function forms the asset's op: it's responsible for producing the asset's contents. The asset in this example doesn't depend on any other assets.
+By default, the name of the decorated function, `my_asset`, is used as the asset key. The decorated function is responsible for producing the asset's contents. The asset in this example doesn't depend on any other assets.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:66]: [1]

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 15]
```diff
@@ -213,22 +218,22 @@ defs = Definitions(
 )
 ```
 
-Using source assets has a few advantages over having the code inside of an asset's op load the data:
+Using source assets has a few advantages over having the code inside of an asset's function load the data without specifying any dependencies:
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:221]: [1]

reworked this sentence in general since "having the code inside an assets' op load the data" kinda undermines the `deps` workflow where the asset's op might do just that. I leaned more into the specifying dependencies vs not specifying, but i'm not sure i totally got the idea across quite right. TODO - return to this sentence and re-work

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 16]
```diff
@@ -213,22 +218,22 @@ defs = Definitions(
...
 
 - **The UI can show asset lineage that includes the source assets**. If different asset definitions in different code locations have the same asset key as a <PyObject object="SourceAsset" /> and both code locations are loaded into the underlying webserver, the UI can represent the asset lineage across the code locations. This can be accomplished using [workspace files](/concepts/code-locations/workspace-files).
 - **Dagster can use data-loading code factored into an <PyObject object="IOManager" /> to load the contents of the source asset**.
-- **Asset dependencies can be written in a consistent way,** independent of whether they're downstream from a source asset or a derived asset. This makes it easy to swap out a source asset for a derived asset and vice versa.
+- **Asset dependencies can be written in a consistent way,** independent of whether they're downstream from a source asset or a software-defined asset. This makes it easy to swap out a source asset for a software-defined asset and vice versa.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:225]: the term "derived asset" isn't used anywhere else in this page and it's kinda confusing what we mean here. Swapped for "software-defined asset" but i'm not sure that's totally accurate either

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 17]
```diff
@@ -213,22 +218,22 @@ defs = Definitions(
...
 
-### Graph-backed assets and multi-assets
+### Writing more computationally complex assets
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:227]: did a rework of this section to avoid `ops` and `graphs` so that, ideally, a new user would be able to understand what `multi_asset` and `graph backed asset` do without needing to first understand other concepts

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 18]
```diff
@@ -306,23 +311,7 @@ def asset_with_version():
...
-        "b": AssetOut(code_version="2"),
-    }
-)
-def multi_asset_with_versions():
-    yield Output(100, "a")
-    yield Output(200, "b")
-```
-
-Just as with regular assets, these versions are attached to the `AssetMaterialization` objects for each of the constituent assets and represented in the UI.
+When an asset with a code version is materialized, it is tagged with the version. The UI will indicate when an asset has a different code version than the code version used for its most recent materialization.
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:314]: removed the `AssetMaterialization` here since this is the first mention of it and i could see that being confusing. Additionally, i don't think it's that important that the user knows we tag the `AssetMaterialization` event (unless their like a power user maybe) but open to push-back on that conclusion

File: docs/content/concepts/assets/software-defined-assets.mdx [hunk 19]
```diff
@@ -615,7 +604,7 @@ def downstream_asset(upstream_asset):
 
 ### Recording materialization metadata
 
-Dagster supports attaching arbitrary [metadata](/\_apidocs/ops#dagster.MetadataValue) to asset materializations. This metadata will be displayed on the "Activity" tab of the "Asset Details" page in the UI. If it's numeric, it will be plotted. To attach metadata, your asset's op can return an <PyObject object="Output" /> object that contains the output value and a dictionary of metadata:
+Dagster supports attaching arbitrary [metadata](/\_apidocs/ops#dagster.MetadataValue) to asset materializations. This metadata will be displayed on the "Activity" tab of the "Asset Details" page in the UI. If it's numeric, it will be plotted. To attach metadata, your asset can return an <PyObject object="Output" /> object that contains the output value and a dictionary of metadata:
```
Comment (jamiedemaria) [inline docs/content/concepts/assets/software-defined-assets.mdx:607]: [1]

**Review by jamiedemaria — COMMENTED** [review]
did another pass with some more updates

As a side note, I definitely feel like I might be over-indexing on the experience of a first time reader, potentially at the expense of an experienced dagster user who is looking for more in-depth information. let me know if you also think that is the case and i can backtrack a bit to find a happy medium. But i do think it's an interesting exercise to re-work this page for a first time user, it's really opened my eyes to the vast number of concepts we toss out at the get-go

File: docs/content/concepts/assets/multi-assets.mdx [hunk 20]
```diff
@@ -109,7 +111,7 @@ def split_actions(context):
 
 Because our outputs are now optional, we can no longer rely on returning a tuple of values, as we don't know in advance which values will be computed. Instead, we explicitly `yield` each output that we're expected to create.
 
-### Dependencies inside multi-assets
+## Dependencies inside multi-assets
```
Comment (erinkcochran87) [inline docs/content/concepts/assets/multi-assets.mdx:114]:
```suggestion
---

## Dependencies inside multi-assets
```

File: docs/content/concepts/assets/multi-assets.mdx [hunk 21]
```diff
@@ -132,3 +134,23 @@ def my_complex_assets(a, b):
     # d only depends on b
     yield Output(value=b + 1, output_name="d")
 ```
+
+---
+
+## Multi-asset code versions
+
+Like regular assets, multi-assets may assign different code versions for each of their outputs. Versions let you help Dagster track what assets haven't been re-materialized since their code has changed, and avoid performing redundant computation.
```
Comment (erinkcochran87) [inline docs/content/concepts/assets/multi-assets.mdx:142]: D'you think we should link to the versioning/caching guide here? https://docs.dagster.io/guides/dagster/asset-versioning-and-caching

I feel like I want *something* more about versioning, but not sure if this is the right page - or if we should maybe create something new that's not as complicated as that guide.

File: docs/content/concepts/assets/software-defined-assets.mdx (hunk 17 above)
Comment (jamiedemaria) [reply docs/content/concepts/assets/software-defined-assets.mdx:227]: that's a good callout. i will think on what this could be re-titled

File: docs/content/concepts/assets/asset-observations.mdx (hunk 8 above)
Comment (jamiedemaria) [reply docs/content/concepts/assets/asset-observations.mdx:8]: tbh me too. I think it's mostly helpful for reporting metadata about an asset after it's been materialized, which can be helpful if your asset calls out to an external tool that handles storing data for you (rather that doing IO managers). However, it looks like a lot of the content in the page is pretty old (like from right around when assets were first introduced), so it might be due for a refresh to account for all the changes since then

**Comment by jamiedemaria** [comment]
I'm in the process of splitting this PR out into several PRs. closing this to help clear up review queues
